client/asset/eth/eth.go

```diff
@@ -3692,7 +3701,16 @@ func (w *assetWallet) withContractor(contractVer uint32, f func(contractor) erro
 	if !found {
 		return fmt.Errorf("no version %d contractor for asset %d", contractVer, w.assetID)
 	}
-	return f(contractor)
+	err := f(contractor)
+	// If a transaction fails with a provider, trust the next nonce.
+	if err != nil {
+		if c, is := contractor.(*contractorV0); is {
+			if mRPC, is := c.cb.(*multiRPCClient); is {
+				mRPC.voidUnusedNonce()
+			}
+		}
+	}
+	return err
```

Comment:
It certainly would be convenient to have the nonce returning done in a more central place like this, but `withContractor` is not just for contract interactions that involve a transaction.  We have read-only contract API requests that involve no transaction and thus no preceding new-nonce request.

I think this needs to be done in the same function that did the nonce reservation itself, i.e. `initiate`, `redeem`, etc.

```go
func (w *assetWallet) initiate(ctx context.Context, assetID uint32, contracts []*asset.Contract,
	maxFeeRate, gasLimit uint64, contractVer uint32) (tx *types.Transaction, err error) {

	var val uint64
	if assetID == BipID {
		for _, c := range contracts {
			val += c.Value
		}
	}
	w.nonceSendMtx.Lock()
	defer w.nonceSendMtx.Unlock()
	// 1. get nonce
	txOpts, _ := w.node.txOpts(ctx, val, gasLimit, dexeth.GweiToWei(maxFeeRate))
	// 2. attempt transaction with the nonce
	return tx, w.withContractor(contractVer, func(c contractor) error {
		tx, err = c.initiate(txOpts, contracts)
		if err != nil {
			// 3. return the unused nonce
			c.voidUnusedNonce()
		}
		return err
	})
}
```

```go
func (c *contractorV0) voidUnusedNonce() {
	if mRPC, is := c.cb.(*multiRPCClient); is {
		mRPC.voidUnusedNonce()
	}
}
```

Even that is a touch fishy though because it requires implementing `voidUnusedNonce`, but maybe alright.

Other ways to do this and be sure we're only targeting failed _transactions_?

Reply:
There could be a `txWithContractor` function that is the same as `withContractor` but also voids the nonce. Then use that function for `initiate`, `redeem`, etc.

Reply:
Put the void in init, redeem, refund, approve, and the sends https://github.com/decred/dcrdex/compare/b20ce45607fe929b82d3127ed59e153f70dcb566..f683573222317c8399e82ccff3439ab5e161f2c4